### PR TITLE
Add benchmarking script with realistic content

### LINKED ISSUE #65: Add benchmarking tests with real-world content
At the moment the `exports` test suite is used to measure performance. While the content in it is significant, it's not really representative of real-world workloads.

Once someone has an open-source site using this, I would be keen to dump all of its content in this repository so we have more relevant numbers to look at.

### LINKED ISSUE #66: Measure memory footprint of the exporter
To make sure we do not cause any memory leak, and this is ready to be used in production servers.

### PR DESCRIPTION
**Pull request by thibaudcolas**
Adds a new `benchmark.py` script that uses a large content sample from https://github.com/thibaudcolas/markov_draftjs (792 ContentStates, ±1000 entities and inline styles). Measures the speed like `example.py` does, as well as the memory consumption.

Fixes #65, fixes #66.

The memory profiling uses a module called `memory_profiler`. It comes with line-by-line profiling of functions (with the `@profile` decorator), as well as a CLI that measures the consumption of the whole script every N seconds and can plot the result.

Here is what is displayed at the moment, in the CI tests:

```
Line #    Mem usage    Increment   Line Contents
================================================
    94     39.1 MiB      0.0 MiB   @profile
    95                             def memory_consumption_run():
    96     39.1 MiB      0.0 MiB       exporter = HTML(config)
    97
    98     39.3 MiB      0.1 MiB       for content_state in content_states:
    99     39.3 MiB      0.0 MiB           exporter.render(content_state)
```

And here is the whole-script measurement + plotting (not integrated ATM):

memory consumption

This is quite cool, but very clunky to use because the charting is based on matplotlib, which can't be installed in a virtual environment. Command I used: `mprof run benchmark.py && deactivate && mprof plot && source .venv/bin/activate` (`memory_profiler` is installed inside and outside the virtual env, `matplotlib` only outside).

### CODE REVIEW

File: .githooks/deploy [hunk 1]
```diff
@@ -2,11 +2,6 @@
 
 # https://gist.github.com/apexskier/efb7c1aaa6e77e8127a8
 # Deploy hooks stored in your git repo to everyone!
-#
-# I keep this in $ROOT/$HOOK_DIR/deploy
-# From the top level of your git repo, run ./hook/deploy (or equivalent) after
-# cloning or adding a new hook.
-# No output is good output.
```
Comment (thibaudcolas) [inline .githooks/deploy]: Unrelated changes to the hooks - just cleanup.

File: benchmark.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import os
+from pstats import Stats
+from memory_profiler import profile
+
+from draftjs_exporter.constants import BLOCK_TYPES, ENTITY_TYPES
+from draftjs_exporter.defaults import BLOCK_MAP, STYLE_MAP
+from draftjs_exporter.dom import DOM
+from draftjs_exporter.html import HTML
+
+from markov_draftjs import get_content_sample
```
Comment (thibaudcolas) [inline benchmark.py]: This content sample is about 2.3Mb of text, which is why it's pulled from a separate package.

File: benchmark.py [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+
+from example import ListItem, OrderedList, Image, BR, EntityFallback
+
+
+def Document(props):
+    return DOM.create_element('a', {
+        'title': props.get('label'),
+        'href': '/documents/%s' % props.get('id'),
+    }, props['children'])
```
Comment (thibaudcolas) [inline benchmark.py]: It's not that important what those decorators do as long as it's realistic with the processing of a real-world integration.

File: example.py [hunk 4]
```diff
@@ -144,519 +144,520 @@ def EntityFallback(props):
...
-        # The most basic mapping format, block type to tag name.
-        BLOCK_TYPES.HEADER_TWO: 'h2',
-        # Use a dict to define props on the block.
-        BLOCK_TYPES.HEADER_THREE: {'element': 'h3', 'props': {'class': 'u-text-center'}},
-        # Add a wrapper (and wrapper_props) to wrap adjacent blocks.
-        BLOCK_TYPES.UNORDERED_LIST_ITEM: {
-            'element': 'li',
-            'wrapper': 'ul',
-            'wrapper_props': {'class': 'bullet-list'},
+if __name__ == '__main__':
```
Comment (thibaudcolas) [inline example.py]: Moving most of the code in here so it's not ran when importing parts of the `example.py` file inside `benchmark.py`.

File: setup.py [hunk 5]
```diff
@@ -25,6 +25,11 @@
     # Required for running the tests.
     'tox>=2.3.1',
 
+    # Benchmark dependencies.
+    'markov_draftjs==0.1.1',
+    'memory-profiler==0.47',
+    'psutil==5.4.1',
```
Comment (thibaudcolas) [inline setup.py]: `psutil` makes `memory_profiler` faster.

**Comment by thibaudcolas** [comment]
There doesn't seem to be much to report on memory-wise so I've merged this right away. Here is the result of a 1000 runs of the exporter on the `markov_draftjs` content sample:

memory-profile-draftjs-exporter-2

I think the red lines mark the time at which memory consumption is at its highest. We can see two things:

- The exporter doesn't leak memory over its execution.
- There is a small (0.1 - 0.5mb) increase during the first ± 200 seconds. I think this comes from the 1...1000 loop but not sure exactly – it does not seem to be present when removing the looping.

**Comment by su27** [comment]
Wow, the `memory_profiler` and the plotting image is so cool!

**Comment by thibaudcolas** [comment]
Some more notes about the process before I completely forget what I did:

Thinking more about it, I'm pretty sure the increase in memory shown on the graph above is because of entities and inline styles in the content. The `markov_draftjs` dataset is ordered with the bigger, more rich text heavy content samples at the start, and the simpler content samples at the end.

The overall memory usage stated above is 39.1 MiB. Of those,

- ± 10 are the base Python process (ie. an empty script consumes that much)
- ± 15 is for the test content.
- ± 8 is for BeautifulSoup.
- That leaves us with ± 7 MiB for the exporter itself.

**Comment by thibaudcolas** [comment]
In the end I wrote a blog post about this! http://blog.thib.me/python-memory-profiling-for-the-draft-js-exporter/
